Mathlib/Data/List/Defs.lean

```diff
@@ -0,0 +1,597 @@
+#align list.of_fn List.ofFn
+#align list.of_fn_nth_val List.ofFnNthVal
+#align list.disjoint List.Disjoint
+#align list.pairwise List.Pairwise
+#align list.pairwise_cons List.pairwise_cons
+#align list.decidable_pairwise List.instDecidablePairwise
+#align list.pw_filter List.pwFilter
+#align list.chain List.Chain
+#align list.chain' List.Chain'
+#align list.chain_cons List.chain_cons
+
+section Chain
+
+noncomputable instance decidableChain [DecidableRel R] (a : α) (l : List α) :
+    Decidable (Chain R a l) := by
+  induction l generalizing a with
+  | nil => simp only [List.Chain.nil]; infer_instance
+  | cons a as ih => haveI := ih; simp only [List.chain_cons]; infer_instance
+#align list.decidable_chain List.decidableChain
+
+noncomputable instance decidableChain' [DecidableRel R] (l : List α) : Decidable (Chain' R l) := by
+  cases l <;> dsimp only [List.Chain'] <;> infer_instance
+#align list.decidable_chain' List.decidableChain'
+
+end Chain
+
+#align list.nodup List.Nodup
+#align list.nodup_decidable List.nodupDecidable
+
+/-- `dedup l` removes duplicates from `l` (taking only the last occurrence).
+  Defined as `pwFilter (≠)`.
+
+     dedup [1, 0, 2, 2, 1] = [0, 2, 1] -/
+def dedup [DecidableEq α] : List α → List α :=
+  pwFilter (· ≠ ·)
+#align list.dedup List.dedup
+
+/-- Greedily create a sublist of `a :: l` such that, for every two adjacent elements `a, b`,
+`R a b` holds. Mostly used with ≠; for example, `destutter' (≠) 1 [2, 2, 1, 1] = [1, 2, 1]`,
+`destutter' (≠) 1, [2, 3, 3] = [1, 2, 3]`, `destutter' (<) 1 [2, 5, 2, 3, 4, 9] = [1, 2, 5, 9]`. -/
+def destutter' (R : α → α → Prop) [DecidableRel R] : α → List α → List α
+  | a, [] => [a]
+  | a, h :: l => if R a h then a :: destutter' R h l else destutter' R a l
+#align list.destutter' List.destutter'
+
+-- TODO: should below be "lazily"?
+/-- Greedily create a sublist of `l` such that, for every two adjacent elements `a, b ∈ l`,
+`R a b` holds. Mostly used with ≠; for example, `destutter (≠) [1, 2, 2, 1, 1] = [1, 2, 1]`,
+`destutter (≠) [1, 2, 3, 3] = [1, 2, 3]`, `destutter (<) [1, 2, 5, 2, 3, 4, 9] = [1, 2, 5, 9]`. -/
+def destutter (R : α → α → Prop) [DecidableRel R] : List α → List α
+  | h :: l => destutter' R h l
+  | [] => []
+#align list.destutter List.destutter
+
+#align list.range' List.range'
+#align list.reduce_option List.reduceOption
+#align list.ilast' List.ilast'
+#align list.last' List.last'
+#align list.rotate List.rotate
+#align list.rotate' List.rotate'
+
+
+section Choose
+
+variable (p : α → Prop) [DecidablePred p] (l : List α)
+
+/-- Given a decidable predicate `p` and a proof of existence of `a ∈ l` such that `p a`,
+choose the first element with this property. This version returns both `a` and proofs
+of `a ∈ l` and `p a`. -/
+def chooseX : ∀ l : List α, ∀ _ : ∃ a, a ∈ l ∧ p a, { a // a ∈ l ∧ p a }
+  | [], hp => False.elim (Exists.elim hp fun a h => not_mem_nil a h.left)
+  | l :: ls, hp =>
+    if pl : p l then ⟨l, ⟨mem_cons.mpr <| Or.inl rfl, pl⟩⟩
```

Comment:
aargh here we are being bitten by the fact that List.mem_cons is not `rfl` like it was in Lean 3 :-( It's not hard to imagine this issue showing up again and again...